### DISCUSSION TITLE
Importing html with images creates note folders, uploading or pasting images does not

### DISCUSSION THREAD

**Discussion by maphew** [TriliumNext/Trilium#6017] [General]
Web pages saved with Chrome or Firefox are as `Foobar.html` and any referenced images, css, js, ... are saved in `Foobar_files` sub-folder. When such a page is wrapped in a zip and imported into Trilium the result is a Note, a Note-folder, and the contents of the _files sub-folder as attachments to Note-folder:

![An imported html file with note-folder and image](https://github.com/user-attachments/assets/33d8be83-fddd-4927-89dd-a5fefca4da33)

If the same note is created interactively in Trilium and using the Insert Image or attachment features the images and attachments are hidden from the Table of Contents view, only the Note is shown.

![Identical note, but only 1 ToC item](https://github.com/user-attachments/assets/8513f83f-fe42-4470-b7b2-cc5f10ac3d51)

From a navigation and node management point of view the interactive result is much preferred -- we only have one item to see and manipulate. It's much less cluttered.

Is there a reason import does not create the same result? Speculation: yes, because any files in the `..._files` folder that are not referred to in the text will be deleted when the automatic database cleanup routines run, we don't want to risk information loss.

Initially this post was going to be a feature request to change import so it replicates the interactive result: a single note with images and attachments hidden from view until you go to _Note >> 3-dot >> Attachments_ menu. However I realised the current Import result has a side effect that is potentially very useful: Since the parent Note-folder owns and references the assets, the clean-up janitor leaves them alone, which allows re-use of images and other attachments without duplication.  

>Each note exclusively owns its attachments, meaning attachments cannot be shared or linked from one note to another. If an attachment link is copied to a different note, the attachment itself is duplicated, and the copies are managed independently thereafter.
>
>Attachments, especially image files, are the recommended method for embedding visuals in notes. It is important to link image attachments within the text of the owning note; otherwise, they will be automatically deleted after a configurable timeout period if not referenced.
>
>https://triliumnext.github.io/Docs/Wiki/attachments.html

So is there a way to have the best of both worlds: a clean and unclutted Table of Contents, and a folder of assets that can be re-used in multiple notes without duplication? 

I think so: after import automatically move `*_files` note-folders into a root level `Assets` or `Media` branch (name configurable), perhaps in the 'Hidden Notes' subtree.

Maybe there are other, smarter approaches though? What do others think?

**Comment by perfectra1n** [comment]
@maphew how are you performing "Web pages saved with Chrome or Firefox"? Is this using the web-clipper or something else? This is most likely an artifact of when Trilium didn't hide pictures or related media to the user (when pictures were their own note).

Would you also be able to just quickly document the steps needed to reproduce what you're seeing? Would help the sanity quite a bit 🤣

**Reply by maphew** [reply]
Example - [Screenshot tour _ Docs_files.zip](https://github.com/user-attachments/files/17544828/Screenshot.tour._.Docs_files.zip) - created by using `Ctrl-S` while viewing a web page in Firefox, then using Import form 3-dots menu on right hand side of a note in Trilium. 

<img width="169" alt="2024-10-27 07_09_06-National Hydrographic Network - TriliumNext Notes" src="https://github.com/user-attachments/assets/eebe3be7-64f1-410b-9062-329b57e75ab6">

The hidden context behind the thread is I have about a 1000 of such pages to import created from Onenote, which uses the same html+files structure as above. (Historically I think it was Internet Explorer that first started this 'format' and FF and Chrome adopted it.) I didn't mention Onenote because not everybody uses it, while most everybody has access to Chrome.

**Reply by perfectra1n** [reply]
Before I take a deeper dive, it looks like @meichthys has done a [deep dive](https://github.com/zadam/trilium/issues/1013#issuecomment-1532108903) on exactly this and their comments might help you @maphew :)

**Reply by meichthys** [reply]
It's been a while since I've done one of these migrations, but I do think the Legacy evernote export path was indeed the best for preserving the formatting and content. @maphew have you tried that? I'm not sure how it compares to the Onemore OneNote addin.

I think importing from a saved browser website may not be a common use case since we have the web-clipper that does that for us quite nicely.

**Reply by maphew** [reply]
@meichthys no I haven't tried the Evernote route. I've never used it so it's a bit low on my prospective tools list, not that it would stop me if that's the best way to go from A to Z! The thought of having to re-link all the internal page links gives me pause. Onemore Archive export adjusts links to work across the .htm files, and those links continue to work after Trilium import. 

That Evernote deep dive is an excellent resource though. Several things there can also be applied to what I'm trying to do.

_also see #444_

**Reply by meichthys** [reply]
That's impressive how the OneMore add-in can preserve page links. How does it handle OneNote links that link to a specific paragraph? 

I'm not sure if the Evernote Legacy client can import an html directory structure, but if it can, maybe you could post-process the OneMore export in Evernote to properly handle the images - and get the best of both worlds?

**Reply by maphew** [reply]
Looks like archive.org might still be having some troubles, I haven't been able to completely download an installer for Evernote Legacy yet. Maybe on the next attempt. 

re: Evernote imports, there are many items in the [Zadam thread](https://github.com/zadam/trilium/issues/1013#issuecomment-1532108903) that didn't make to [the docs](https://triliumnext.github.io/Docs/Wiki/evernote-import.html) 😉

**Reply by maphew** [reply]
> Looks like archive.org might still be having some troubles, I haven't been able to completely download an installer for Evernote Legacy yet. Maybe on the next attempt.

hah, Cunningham's Law in action. No sooner claimed than proved wrong! I have downloaded and installed v6.5. It won't start though - complains with _"No connection to server. You need to be online to register or sync to server."_ with the only option being to quit or try again. I am online, so maybe they've cut old version access.

...going back to self-mangling folders. Perhaps a good excuse to extend my slowly growing _trilium-py_ skills. :)

**Reply by meichthys** [reply]
> re: Evernote imports, there are many items in the [Zadam thread](https://github.com/zadam/trilium/issues/1013#issuecomment-1532108903) that didn't make to [the docs](https://triliumnext.github.io/Docs/Wiki/evernote-import.html) 😉

See here: https://triliumnext.github.io/Docs/Wiki/onenote.html

> Looks like archive.org might still be having some troubles, I haven't been able to completely download an installer for Evernote Legacy yet. Maybe on the next attempt.

Were you using the download from here? https://web.archive.org/web/20230327110646/https://help.evernote.com/hc/en-us/articles/360052560314
I never had to sign in or sync afaik. Perhaps you could create a temp account and then just disconnect your internet connection while doing the import/export. I'm pretty sure I didn't need to do that though.

**Reply by maphew** [reply]
Thanks, I see those extra bits on the Onenote page now.

And yes, I'm using the web archive download. There's no option given to create an account. The application doesn't appear at all. On launch there's the server connection error, of which the only options are Try Again and Quit. Try again just repeats the window, and quit, well, it quits. :)

**Reply by perfectra1n** [reply]
Even when disabling the network interfaces on your machine? That’s a bummer

**Comment by maphew** [comment]
> ...after import automatically move *_files note-folders into a root level Assets or Media branch (name configurable), perhaps in the 'Hidden Notes' subtree.

This works, but raises challenges for notes that will be shared: if the Assets/Media branch is not shared then the referenced media files will not display. So for this method to work there needs to be 2 Media notes: 1 for private, the default, and one for public. The operator needs to be aware and move media from one to the other -- and back if un-sharing.

So while it's possible I wouldn't recommend. Over the course of time a person _will_ forget to manage the media files and make files unreachable, or worse, public when they're meant to be private.

The better solution is to change html import so it embeds assets as if they were pasted in instead of referencing them.
